pkg/controller/common.go

```diff
@@ -788,18 +832,108 @@ func getNodeRemainingCapacity(ctx context.Context, cachedClient client.Client, n
 	return capacity - len(attachments), nil
 }
 
-func selectNodes(ctx context.Context, azr azReconciler, numNodes int, primaryNode string, exceptionNodes map[string]int, podNodeAffinities []*nodeaffinity.RequiredNodeAffinity, podTolerations []v1.Toleration, podNodeSelectors, podAffinities labels.Selector, volumeNodeSelectors []*nodeaffinity.NodeSelector) ([]string, error) {
-	filteredNodes := []string{}
+func selectNodes(ctx context.Context, azr azReconciler, numNodes int, primaryNode string, exceptionNodes map[string]int, podNodeAffinities []*nodeaffinity.RequiredNodeAffinity, podTolerations []v1.Toleration, podNodeSelectors, podAffinities labels.Selector, volumeNodeSelectors []*nodeaffinity.NodeSelector, compatibleZones []string) ([]string, error) {
+	selectedNodes := []string{}
 	var nodes *v1.NodeList
 	var err error
 
-	// List all nodes
 	nodes = &v1.NodeList{}
+	// Select only the nodes from specific supported zones in case of a multi zone cluster
+	if len(compatibleZones) > 0 {
+		klog.Infof("The list of zones to select nodes from is: %s", strings.Join(compatibleZones, ","))
+		nodeObj := &v1.Node{}
+		// Get the zone of the primary node
+		err := azr.getClient().Get(ctx, types.NamespacedName{Name: primaryNode}, nodeObj)
+		if err != nil {
+			klog.Errorf("failed to retrieve the primary node: %v", err)
+		}
+		primaryNodeZone := nodeObj.Labels[consts.AzureDiskCSITopologyKey]
```

Comment:
We should be defensive and handle the case where the driver tolopogy key is not present.